e2e/e2e_openssl_test.go

```diff
@@ -0,0 +1,242 @@
+// +build openssl,!js
+
+package e2e
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/pion/dtls/v2"
+)
+
+func serverOpenSSL(c *comm) {
+	go func() {
+		c.serverMutex.Lock()
+		defer c.serverMutex.Unlock()
+
+		cfg := c.serverConfig
+
+		// create openssl arguments
+		args := []string{"s_server",
+			"-dtls1_2",
+			"-quiet",
+			fmt.Sprintf("-accept=%d", c.serverPort),
+		}
+		ciphers := ciphersOpenSSL(cfg)
+		if ciphers != "" {
+			args = append(args, fmt.Sprintf("-cipher=%s", ciphers))
+		}
+
+		// psk arguments
+		if cfg.PSK != nil {
+			psk, err := cfg.PSK(nil)
+			if err != nil {
+				c.errChan <- err
+				return
+			}
+			args = append(args, fmt.Sprintf("-psk=%X", psk))
+			if len(cfg.PSKIdentityHint) > 0 {
+				args = append(args, fmt.Sprintf("-psk_hint=%s", cfg.PSKIdentityHint))
+			}
+		}
+
+		// certs arguments
+		if len(cfg.Certificates) > 0 {
+			// create temporary cert files
+			certPEM, keyPEM, err := writeTempPEM(cfg)
+			if err != nil {
+				c.errChan <- err
+				return
+			}
+			args = append(args,
+				fmt.Sprintf("-cert=%s", certPEM),
+				fmt.Sprintf("-key=%s", keyPEM))
+			defer func() {
+				_ = os.Remove(certPEM)
+				_ = os.Remove(keyPEM)
+			}()
+		} else {
+			args = append(args, "-nocert")
+		}
+
+		// launch command
+		// #nosec G204
+		cmd := exec.CommandContext(c.ctx, "openssl", args...)
+		var inner net.Conn
+		inner, c.serverConn = net.Pipe()
+		cmd.Stdin = inner
+		cmd.Stdout = inner
+		cmd.Stderr = os.Stderr
+		if err := cmd.Start(); err != nil {
+			c.errChan <- err
+			_ = inner.Close()
+			return
+		}
+
+		// Ensure that server has started
+		time.Sleep(500 * time.Millisecond)
+
+		c.serverReady <- struct{}{}
+		simpleReadWrite(c.errChan, c.serverChan, c.serverConn, c.messageRecvCount)
+	}()
+}
+
+func clientOpenSSL(c *comm) {
+	select {
+	case <-c.serverReady:
+		// OK
+	case <-time.After(time.Second):
+		c.errChan <- errors.New("waiting on serverReady err: timeout")
+	}
+
+	c.clientMutex.Lock()
+	defer c.clientMutex.Unlock()
+
+	cfg := c.clientConfig
+
+	// create openssl arguments
+	args := []string{"s_client",
+		"-dtls1_2",
+		"-quiet",
```

Comment:
Now, master branch supports SNI,
```suggestion
		"-quiet",
		"-servername=localhost",
```
```diff
diff --git a/e2e/e2e_test.go b/e2e/e2e_test.go
index d103f3f..531402b 100644
--- a/e2e/e2e_test.go
+++ b/e2e/e2e_test.go
@@ -234,7 +234,7 @@ func testPionE2ESimple(t *testing.T, server, client func(*comm)) {
                        ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
                        defer cancel()
 
-                       cert, err := selfsign.GenerateSelfSigned()
+                       cert, err := selfsign.GenerateSelfSignedWithDNS("localhost")
                        if err != nil {
                                t.Fatal(err)
                        }
@@ -301,7 +301,7 @@ func testPionE2EMTUs(t *testing.T, server, client func(*comm)) {
                        ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
                        defer cancel()
 
-                       cert, err := selfsign.GenerateSelfSigned()
+                       cert, err := selfsign.GenerateSelfSignedWithDNS("localhost")
                        if err != nil {
                                t.Fatal(err)
                        }

```
will wipe all errors away except verification errors of self-signed certificates.

Reply:
Done!